source/auth/auth.rst

```diff
@@ -1175,70 +1175,146 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
 .. note::
     Five minutes was chosen based on the AWS documentation for `IAM roles for EC2 <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html>`_ : "We make new credentials available at least five minutes before the expiration of the old credentials". The intent is to have some buffer between when the driver fetches the credentials and when the server verifies them.
 
+
 MONGODB-OIDC
 ~~~~~~~~~~~~
 
 :since: 7.0 Enterprise
 
-MONGODB-OIDC authenticates using an `OIDC <https://openid.net/specs/openid-connect-core-1_0.html>`_ access tokens.  Drivers MUST support
-both Callback-driven OIDC and Automatic OIDC Authentication for AWS.
+MONGODB-OIDC authenticates using an `OpenID Connect (OIDC)
+<https://openid.net/specs/openid-connect-core-1_0.html>`_ access token.
+
+TODO: Add background on what part of the auth flow drivers play specifically.
+
+`MongoCredential`_ Properties
+`````````````````````````````
+
+username
+    MUST NOT be specified.
+
+source
+    MUST be "$external". Defaults to ``$external``.
+
+password
+    MUST NOT be specified.
 
+mechanism
+    MUST be "MONGODB-OIDC"
+
+mechanism_properties
+    PROVIDER_NAME
+        Drivers MUST allow the user to specify the name of the OIDC service to
+        use to obtain credentials. MUST be one of ["aws"]. If PROVIDER_NAME is
+        given and a Custom provider callback is configured, the driver MUST
+        raise an error.
+
+Supported Service Providers
+```````````````````````````
+
+AWS
+___
+
+TODO: Add.
+
+Custom
+______
+
+Drivers MUST allow the user to integrate with other OIDC providers by
+implementing a custom callback that returns an OIDC token. Drivers MUST provide
+a way for the callback to be either automatically canceled, or to cancel itself.
+This can be as a timeout argument to the callback, a cancellation context passed
+to the callback, or some other language-appropriate mechanism.  The timeout
+deadline MUST be the same as the connection establishment deadline.
+
+Callbacks can be synchronous and/or asynchronous, depending on the driver
+and/or language.  Asynchronous callbacks should be preferred when other
+operations in the driver use asynchronous functions.
+
+The driver MUST pass the following information to the callback:
+
+- ``timeoutMS``: A timeout, in milliseconds, deadline, or ``timeoutContext``.
+- ``mechanismProperties``: All key-value pairs from the
+  AUTH_MECHANISM_PROPERTIES parameter.
+- ``invalidatedToken``: Optional. The previous token that was invalidated after
+  receiving a ``ReauthenticationRequired`` error.
+
+The callback MUST return an OIDC access token in JWT format.
```

Comment:
Is this intended to be a documentation requirement? (Drivers could enforce this by introducing a type that validates user input, but I believe we were fine with returning strings.)

Reply:
Generally speaking, it’s strongly recommended for clients to not inspect access tokens themselves, only for resource servers. For example, at some theoretical point in the future, MongoDB servers could add support for opaque access tokens, which would be incompatible with added JWT validation in the driver.

Reply:
This was intended to be a documentation requirement to inform users how to implement the callback. I will update it to clarify that.

Reply:
Rewrote this section to only describe what the callback API must support.